src/Diagnostics.DataProviders/DataProviders.cs

```diff
@@ -18,27 +20,85 @@ public class DataProviders
         public Func<MdmDataSource, IMdmDataProvider> Mdm;
         public Func<GenericMdmDataProviderConfiguration, IMdmDataProvider> MdmGeneric;
 
+        private readonly Dictionary<object, IMetadataProvider> _dataProvidersCache = new Dictionary<object, IMetadataProvider>();
```

Comment:
Puneet, maybe we can call this data provider list since its not really a "cache". `List<LogDecoratorBase>`